### PR TITLE
feat(synchro): #MEX-395 adapt api endpoint to return more informations about users to sync

### PR DESCRIPTION
**Pull request by Glama107**
### PR Description

Adaptation of api endpoint, to add more informations about users to sync with PMB.

**Infos added:**

- **Email :**        Personal email of current user
- **Address :**   Road name and number
- **Zip code :**  Postal code
- **Country :**   Current country of user
- **City :**         City 🥇 

**Example of output for one random user :** 

```
 {
        "externalId": "99999999",
        "lastName": "Doe",
        "firstName": "John",
        "login": "john.doe",
        "email": "johndoe@doe.com",
        "emailAcademy": null,
        "mobile": null,
        "deleteDate": null,
        "functions": null,
        "displayName": "DOE John",
        "profiles": "Student",
        "structures": "8888",
        "classes": [
            "777777"
        ],
        "address": "99 Rue Dupont",
        "zipCode": "999999",
        "country": "FRANCE",
        "city": "DoeCity"
    },
``` 

## Issue ticket number and link

[MEX-395](https://jira.support-ent.fr/browse/MEX-395)

## Checklist before requesting a review (magic string, indentation, comment/documentation...)

- [ ] I have detailed the tests to do in my feature/fix in order to prevent consequents regressions (must specify in **Checklist tests**)
- [x] I have performed a self-review of my code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation (API Doc etc...) - (must specify in **Description** for target version)
- [ ] If it is a consequent feature, I have added thorough tests.
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] Any dependent changes have been added to this project (must specify in **Description**)

### CODE REVIEW

File: src/main/java/fr/openent/pmb/services/impl/DefaultUserService.java [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+                        usersInfos.stream()
+                                .filter(infos -> externalId.equals(((JsonObject) infos).getString("externalId")))
+                                .findFirst()
+                                .ifPresent(infoObject -> {
+                                    JsonObject info = (JsonObject) infoObject;
+                                    userObject.put("email", info.getString("email"));
+                                    userObject.put("address", info.getString("address"));
+                                    userObject.put("zipCode", info.getString("zipCode"));
+                                    userObject.put("country", info.getString("country"));
+                                    userObject.put("city", info.getString("city"));
```
Comment (Mattox0) [inline src/main/java/fr/openent/pmb/services/impl/DefaultUserService.java]: ça serait bien de mettre des magic strings :)

File: src/main/java/fr/openent/pmb/services/impl/DefaultUserService.java [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+import io.vertx.core.json.JsonArray;
+import io.vertx.core.json.JsonObject;
+import org.entcore.common.neo4j.Neo4j;
+import org.entcore.common.neo4j.Neo4jResult;
+
+import static io.vertx.core.http.impl.HttpClientConnection.log;
+
+public class DefaultUserService implements UserService {
+
+    public Future<JsonArray> getUsersByStructure(JsonArray UAIs, String type){
```
Comment (Alex-Delaveau) [inline src/main/java/fr/openent/pmb/services/impl/DefaultUserService.java]: Si tu sais que ça va être utilisé que dans ce fichier tu peux passer la visibilité de la fonction à private

File: src/main/java/fr/openent/pmb/services/impl/DefaultUserService.java [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+                "u.zipCode as zipCode,\n" +
+                "u.country as country,\n" +
+                "u.city as city;";
+        JsonObject params = new JsonObject()
+                .put("uais", UAIs)
+                .put("type", type);
+
+        Neo4j.getInstance().execute(query, params, Neo4jResult.validResultHandler(event -> {
+            if (event.isLeft()) {
+                log.error("Pmb-Connector@DefaultUserService::getUsersByStructure Error when trying to execute query");
```
Comment (Mattox0) [inline src/main/java/fr/openent/pmb/services/impl/DefaultUserService.java]: met des [ ] pour que ce soit + iso à toutes les plateformes

File: src/main/java/fr/openent/pmb/services/impl/DefaultUserService.java [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+                                    userObject.put("address", info.getString("address"));
+                                    userObject.put("zipCode", info.getString("zipCode"));
+                                    userObject.put("country", info.getString("country"));
+                                    userObject.put("city", info.getString("city"));
+                                });
+                    });
+                    promise.complete(users);
+                })
+                .onFailure( err -> {
+                    log.error("Pmb-Connector@DefaultUserService::mergeUserInfos Error when trying to get users by structure", err);
```
Comment (Mattox0) [inline src/main/java/fr/openent/pmb/services/impl/DefaultUserService.java]: same [ ]

Comment (Mattox0) [reply src/main/java/fr/openent/pmb/services/impl/DefaultUserService.java]: + ne pas mettre de virugle 
+ err.getMessage()

**Review by Mattox0 — CHANGES_REQUESTED** [review]

File: src/main/java/fr/openent/pmb/controllers/PmbController.java [hunk 5]
```diff
@@ -117,42 +120,40 @@ private Future<List<JsonObject>> addUaisToUsers(JsonArray users, JsonArray users
     public void listUserInStructuresByUAI(final HttpServerRequest request) {
         String uai = request.getParam("uai");
         String type = request.getParam("type");
-        if(uai == null){
-            log.error("No uai sent in request pmb/structures/list");
+        if(uai == null || type == null){
+            log.error("Pmb-connector@PmbController::listUserInStructuresByUAI No uai/type sent in request.");
```
Comment (Mattox0) [inline src/main/java/fr/openent/pmb/controllers/PmbController.java]: same [ ]

**Review by Alex-Delaveau — CHANGES_REQUESTED** [review]

File: src/main/java/fr/openent/pmb/services/impl/DefaultUserService.java (hunk 3 above)
Comment (Alex-Delaveau) [reply src/main/java/fr/openent/pmb/services/impl/DefaultUserService.java]: Par ex startPromise.tryFail("[Competences@Competences::start] Fail to start Competences");

File: src/main/java/fr/openent/pmb/services/impl/DefaultUserService.java (hunk 2 above)
Comment (Glama107) [reply src/main/java/fr/openent/pmb/services/impl/DefaultUserService.java]: Bien vu bogoss

File: src/main/java/fr/openent/pmb/services/impl/DefaultUserService.java (hunk 4 above)
Comment (Glama107) [reply src/main/java/fr/openent/pmb/services/impl/DefaultUserService.java]: Ouais pas mal

File: src/main/java/fr/openent/pmb/services/impl/DefaultUserService.java (hunk 3 above)
Comment (Glama107) [reply src/main/java/fr/openent/pmb/services/impl/DefaultUserService.java]: Merci pour l'exemple c'est carré le boss

**Review by Mattox0 — APPROVED** [review]

**Review by Alex-Delaveau — APPROVED** [review]

File: src/main/java/fr/openent/pmb/controllers/PmbController.java [hunk 6]
```diff
@@ -117,42 +120,40 @@ private Future<List<JsonObject>> addUaisToUsers(JsonArray users, JsonArray users
...
         String uai = request.getParam("uai");
         String type = request.getParam("type");
-        if(uai == null){
-            log.error("No uai sent in request pmb/structures/list");
+        if(uai == null || type == null){
+            log.error("[Pmb-connector@PmbController::listUserInStructuresByUAI] No uai/type sent in request.");
             badRequest(request);
         }else{
             JsonArray UAIs = new JsonArray().add(uai);
-            if (type != null && type.equals(Pmb.MANAGER_TYPE_PRET)) {
```
Comment (JinDouang) [inline src/main/java/fr/openent/pmb/controllers/PmbController.java:125]: normal qu'on enlève la vérification pour manager type pret ?

File: src/main/java/fr/openent/pmb/controllers/PmbController.java [hunk 7]
```diff
@@ -117,42 +120,40 @@ private Future<List<JsonObject>> addUaisToUsers(JsonArray users, JsonArray users
...
+                badRequest(request);
             }
         }
     }
 
-    private void directoryUsersExport(HttpServerRequest request, JsonArray UAIs) {
+    private void directoryUsersExport(HttpServerRequest request, JsonArray UAIs, String type) {
         StringBuilder url = new StringBuilder().append(pmbConfig.getString("host"))
-                .append("/directory/user/structures/list?");
+                .append("/directory/user/structures/list?full=true&");
```
Comment (JinDouang) [inline src/main/java/fr/openent/pmb/controllers/PmbController.java:139]: à quoi sert le full=true ?

**Review by JinDouang — CHANGES_REQUESTED** [review]

File: src/main/java/fr/openent/pmb/controllers/PmbController.java (hunk 6 above)
Comment (Glama107) [reply src/main/java/fr/openent/pmb/controllers/PmbController.java:125]: yep c'est plus le même type qu'on vérifie (Student / Teacher / Personnel)

File: src/main/java/fr/openent/pmb/controllers/PmbController.java (hunk 7 above)
Comment (Glama107) [reply src/main/java/fr/openent/pmb/controllers/PmbController.java:139]: API directory de l'ent-core :
```
//user's fields for Full Export
		if(isExportFull){
			fields.add("email");
			fields.add("emailAcademy");
			fields.add("mobile");
			fields.add("deleteDate");
			fields.add("functions");
			fields.add("displayName");
		}
```

**Review by JinDouang — APPROVED** [review]

**Pull request merged**